### PR TITLE
Feat subtitle support

### PR DESCRIPTION
**Pull request by Sgtfishtank**
Before the work was done in this PR, we did not have any support for subtitles in Channel engine. This pr will solve that by implementing support for WebVTT subtitles.

The work that has been done is to implement the communication between CE and hls-Vod2Live for subtitles, implement endpoints for getting subtitle manifest, handle splicing WebVtt files to match video segment length, and supply empty/dummy WebVtt files.

This PR aims to solve issue https://github.com/Eyevinn/channel-engine/issues/44

### CODE REVIEW

**Comment by birme** [comment]
Adding a reminder that this branch needs to be rebased with `v4_1-rc`

**Comment by birme** [comment]
Changed base to v4_2-rc branch

File: engine/server.ts [hunk 1]
```diff
@@ -185,6 +200,25 @@ export class ChannelEngine {
...
+    } else {
+      options.dummySubtitleEndpoint = this.dummySubtitleEndpoint;
+    }
+    this.subtitleSliceEndpoint = subtitleSliceEndpointPath;
+    if (options && options.subtitleSliceEndpoint) {
+      this.subtitleSliceEndpoint = options.subtitleSliceEndpoint;
+    } else {
+      options.subtitleSliceEndpoint = this.subtitleSliceEndpoint;
+    }
+    this.shouldContainSubtitles = true;
```
Comment (birme) [inline engine/server.ts:215]: Default should be false so we don't break current implementations. Also rename option to be called `useVTTSubtitles`

File: engine/server.ts [hunk 2]
```diff
@@ -185,6 +200,25 @@ export class ChannelEngine {
     if (options && options.useDemuxedAudio) {
       this.useDemuxedAudio = true;
     }
+    this.dummySubtitleEndpoint = dummySubtitleEndpointPath;
```
Comment (birme) [inline engine/server.ts:203]: I am not really sure I understand what you try to accomplish here below. Don't understand the logic here, can this be simplified? Rewrite this to not overwrite input parameters

File: engine/server.ts [hunk 3]
```diff
@@ -709,6 +788,71 @@ export class ChannelEngine {
...
+    }
+  }
+
+  async _handleSubtitleSliceEndpoint(req,res,next) {
+    debug(`req.url=${req.url}`);
+      try {
+        const slicer = new SubtitleSlicer();
+        const body = await slicer.generateVtt(req.query);
+        res.sendRaw(200, Buffer.from(body, 'utf8'), {
+          "Content-Type": "application/x-mpegURL;charset=UTF-8",
```
Comment (birme) [inline engine/server.ts:845]: use the content type `application/vnd.apple.mpegurl`

File: engine/session.js [hunk 4]
```diff
@@ -85,6 +90,15 @@ class Session {
       if (config.useDemuxedAudio) {
         this.use_demuxed_audio = true;
       }
+      if (config.dummySubtitleEndpoint) {
+        this.dummySubtitleEndpoint = config.dummySubtitleEndpoint;
+      }
+      if (config.subtitleSliceEndpoint) {
+        this.subtitleSliceEndpoint = config.subtitleSliceEndpoint;
+      }
+      if (config.shouldContainSubtitles) {
```
Comment (birme) [inline engine/session.js:99]: change `config.shouldContainSubtitles` to `config.useVTTSubtitles`

**Review by birme — CHANGES_REQUESTED** [review]

File: engine/server.ts [hunk 5]
```diff
@@ -709,6 +788,71 @@ export class ChannelEngine {
...
+  }
+
+  async _handleDummySubtitleEndpoint(req,res,next) {
+    debug(`req.url=${req.url}`);
+    const session = sessions[req.params.channelId];
+    if (session) {
+      try {
+        const body = `WEBVTT`;
+        res.sendRaw(200, Buffer.from(body, 'utf8'), {
+          "Content-Type": "application/x-mpegURL;charset=UTF-8",
```
[the remark this answers is not in the document]
Comment (Sgtfishtank) [reply engine/server.ts]: fixed

File: engine/server.ts (hunk 2 above)
Comment (Sgtfishtank) [reply engine/server.ts]: fixed

File: engine/server.ts (hunk 3 above)
Comment (Sgtfishtank) [reply engine/server.ts]: fixed

File: engine/session.js [hunk 6]
```diff
@@ -85,6 +90,15 @@ class Session {
...
+        this.shouldContainSubtitles = config.shouldContainSubtitles;
```
[the remark this answers is not in the document]
Comment (Sgtfishtank) [reply engine/session.js]: fixed

File: engine/session.js (hunk 4 above)
Comment (Sgtfishtank) [reply engine/session.js]: fixed

File: engine/session.js [hunk 7]
```diff
@@ -40,6 +40,9 @@ class Session {
     this._events = [];
     this.averageSegmentDuration = AVERAGE_SEGMENT_DURATION;
     this.use_demuxed_audio = false;
+    this.shouldContainSubtitles = false;
```
[the remark this answers is not in the document]
Comment (Sgtfishtank) [reply engine/session.js]: fixed

File: engine/server.ts (hunk 1 above)
Comment (Sgtfishtank) [reply engine/server.ts]: fixed

File: engine/session.js [hunk 8]
```diff
@@ -40,6 +40,9 @@ class Session {
     this._events = [];
     this.averageSegmentDuration = AVERAGE_SEGMENT_DURATION;
     this.use_demuxed_audio = false;
+    this.use_vtt_Subtitles = false;
```
Comment (birme) [inline engine/session.js:43]: Why the capital `S` here? Would be more consistent to call it `use_vtt_subtitles`

File: server.ts [hunk 9]
```diff
@@ -19,17 +19,17 @@ class RefAssetManager implements IAssetManager {
         {
           id: 1,
           title: "Tears of Steel",
-          uri: "https://maitv-vod.lab.eyevinn.technology/tearsofsteel_4k.mov/master.m3u8",
+          uri: "http://localhost:80/sintel_1min_stream/index.m3u8",
```
Comment (birme) [inline server.ts:22]: Do not refer to local test assets in the reference implementations

**Review by birme — CHANGES_REQUESTED** [review]
Remove the `.DS_Store` and `.vscode/settings.json`
Fix the spelling error in the commit message and cleanup the commit with a more descriptive message and not just a list of commit messages...

File: engine/session.js (hunk 8 above)
Comment (Sgtfishtank) [reply engine/session.js]: fixed

File: server.ts (hunk 9 above)
Comment (Sgtfishtank) [reply server.ts]: reverted back to the original

**Pull request merged**